### PR TITLE
backport: merge bitcoin#21148, #21327, #23970, #24021, #24543, #26844, #25325, #28165, partial bitcoin#20524, #26036, #27981 (networking backports: part 7)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* Dependent on https://github.com/dashpay/dash/pull/6098

* Dependent on https://github.com/dashpay/dash/pull/6233

* `p2p_ibd_txrelay.py` was first introduced in [bitcoin#19423](https://github.com/bitcoin/bitcoin/pull/19423) to test feefilter logic but on account of Dash not having feefilter capabilities, that backport was skipped over but on account of the tests introduced in [bitcoin#21327](https://github.com/bitcoin/bitcoin/pull/21327) that test capabilities present in Dash, a minimal version of `p2p_ibd_txrelay.py` has been committed in.

* `vSendMsg` is originally a `std::deque` and as an optimization, was changed to a `std::list` in 027a852a ([dash#3398](https://github.com/dashpay/dash/pull/3398)) but this renders us unable to backport [bitcoin#26844](https://github.com/bitcoin/bitcoin/pull/26844) as it introduces build failures. The optimization has been reverted to make way for the backport.

  <details>

  <summary>Compile failure:</summary>

  ```
  net.cpp:959:20: error: invalid operands to binary expression ('iterator' (aka '_List_iterator<std::vector<unsigned char, std::allocator<unsigned char>>>') and 'int')
              if (it + 1 != node.vSendMsg.end()) {
                  ~~ ^ ~
  /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_bvector.h:303:3: note: candidate function not viable: no known conversion from 'iterator' (aka '_List_iterator<std::vector<unsigned char, std::allocator<unsigned char>>>') to 'ptrdiff_t' (aka 'long') for 1st argument
    operator+(ptrdiff_t __n, const _Bit_iterator& __x)
  [...]
  1 error generated.
  make[2]: *** [Makefile:11296: libbitcoin_server_a-net.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
  make[2]: Leaving directory '/src/dash/src'
  make[1]: *** [Makefile:19171: all-recursive] Error 1
  make[1]: Leaving directory '/src/dash/src'
  make: *** [Makefile:799: all-recursive] Error 1
  ```

  </details>

* The collection of `CNode` pointers in `CConnman::SocketHandlerConnected` has been changed to a `std::set` to allow for us to erase elements from `vReceivableNodes` if the node is _also_ in the set of sendable nodes and the send hasn't entirely succeeded to avoid a deadlock (i.e. backport [bitcoin#27981](https://github.com/bitcoin/bitcoin/pull/27981)) 

* When backporting [bitcoin#28165](https://github.com/bitcoin/bitcoin/pull/28165), `denialofservice_tests` has been modified to still check with `vSendMsg` instead of `Transport::GetBytesToSend()` as changes in networking code to support LT-SEMs (level-triggered socket events mode) mean that the message doesn't get shifted from `vSendMsg` to `m_message_to_send`, as the test expects.
  * Specifically, the changes made for LT-SEM support result in the function responsible for making that shift (`Transport::SetMessageToSend()` through `CConnman::SocketSendData()`), not being called during the test runtime.

* As checking `vSendMsg` (directly or through `nSendMsgSize`) isn't enough to determine if the queue is empty, we now also check with `to_send` from `Transport::GetBytesToSend()` to help us make that determination. This mirrors the change present in the upstream backport ([source](https://github.com/bitcoin/bitcoin/pull/28165/files#diff-00021eed586a482abdb09d6cdada1d90115abe988a91421851960e26658bed02R1324-R1327)). 

## Breaking Changes

* `bandwidth.message.*.bytesSent` will no longer include overhead and will now only report message size as specifics that let us calculate the overhead have been abstracted away.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -1715,10 +1715,10 @@ bool CConnman::GenerateSelectSet(const std::vector<CNode*>& nodes,
         recv_set.insert(hListenSocket.sock->Get());
     }
 
-    for (CNode* pnode : nodes)
-    {
+    for (CNode* pnode : nodes) {
         bool select_recv = !pnode->fHasRecvData;
         bool select_send = !pnode->fCanSendData;
+        if (!select_recv && !select_send) continue;
```
Comment (UdjinM6) [inline src/net.cpp:1721]: 27981: our net logic is different, pretty sure we should drop this line

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/test_framework/test_node.py [hunk 2]
```diff
@@ -594,7 +594,7 @@ def addconnection_callback(address, port):
 
     def num_test_p2p_connections(self):
         """Return number of test framework p2p connections to the node."""
-        return len([peer for peer in self.getpeerinfo() if peer['subver'] == MY_SUBVERSION])
+        return len([peer for peer in self.getpeerinfo() if peer['subver'] == P2P_SUBVERSION])
```
[the remark this answers is not in the document]
Comment (kwvg) [reply test/functional/test_framework/test_node.py]: Resolved in latest push

File: src/net.cpp (hunk 1 above)
Comment (kwvg) [reply src/net.cpp]: Resolved in latest push

**Review by UdjinM6 — DISMISSED** [review]
Seems to be working with no issues.

light ACK 71a5e90b3b7c00dbf89248caf1c58f2835cfad78

**Review by PastaPastaPasta — DISMISSED** [review]
generally LGTM: 71a5e90b3b7c00dbf89248caf1c58f2835cfad78; needs clang-format fixed

**Review by UdjinM6 — DISMISSED** [review]
utACK d0faa3aaf194ccb1b67ea32f29cb4772f5068c50

**Review by PastaPastaPasta — APPROVED** [review]
utACK 76a458e5f947e92376ecef17f2b1eef084247701

**Pull request merged**